### PR TITLE
LUCENE-9047: Move the Directory APIs to be little endian (take 2)

### PR DESCRIPTION
**Pull request by iverase**
Here is a proposal for changing the directory API to be little endian meanwhile keeping backwards compatibility. This effort build on top of the efforts of having new codecs for Lucene 9.0. In order to illustrate the approach, the PR is divided in 5 commits. Here is the explanation for each commit:

**commit 1**: When changing the Directory API, the ways we write and read integers from DirectReader / DirectWriter needs to change as well. Therefore we need to make a copy of those classes to backwards codecs and make sure those codecs are using that version instead of the one in core.  This applies as well to DirectMonotonicReader / DirectMonotonicWriter.

**commit 2**: I am still proposing to use IndexInput / IndexOutput wrappers to handle backwards compatibility. In order to make it easier to read, this commit wraps all directory calls in backward codecs that create an IndexInput / IndexOutput. 

**commit 3**:  There is one file that we need to open that it does not belong to a codec and it might be written in big endian or little endian. The file is `segment.gen`. In order to make it easy, I took the approach to write this file always using big endian encoding. In addition I am doing the same for codec headers / footers. This can be improved but this approach helped me moving forward at this point.

**commit 4**:  This commit actually changes the Directory endianness  and updates the DirectReader / DirectWriter integer packers to work with this endianness. It introduces the IndexInput / IndexOutput wrappers for backwards codecs.

**commit 5**: Last changes to fix the last failing tests. In particular adds backwards compatibility for FST.

The idea with this PR is to agree in the procedure. If we can agree my proposal is to. add commit 1 and 2 first as they are only refactors. Then we can focus in the last commits.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/codecs/CodecUtil.java [hunk 1]
```diff
@@ -640,6 +640,33 @@ static void writeCRC(IndexOutput output) throws IOException {
       throw new IllegalStateException(
           "Illegal CRC-32 checksum: " + value + " (resource=" + output + ")");
     }
-    output.writeLong(value);
+    writeLong(output, value);
+  }
+
+  /** write int value on header / footer */
+  public static void writeInt(DataOutput out, int i) throws IOException {
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/codecs/CodecUtil.java:647]: Maybe say explicitly on these methods that they write in big-endian order?

File: lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/store/DirectoryUtil.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+
+/**
+ * Utility class to wrap open files
+ *
+ * @lucene.internal
+ */
+public final class DirectoryUtil {
```
Comment (jpountz) [inline lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/store/DirectoryUtil.java:33]: Give it a more descriptive name, e.g. `EndiannessReverserUtil` or something along these lines for consistency with the input/output wrapper names?

**Review by jpountz — APPROVED** [review]
+1 to the approach, the change looks great to me

File: lucene/core/src/java/org/apache/lucene/store/ByteBufferIndexInput.java [hunk 3]
```diff
@@ -177,7 +177,7 @@ public void readLELongs(long[] dst, int offset, int length) throws IOException {
   }
 
   @Override
-  public final void readLEFloats(float[] floats, int offset, int len) throws IOException {
+  public final void readFloats(float[] floats, int offset, int len) throws IOException {
     // See notes about readELongs above
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/store/ByteBufferIndexInput.java:181]:
```suggestion
    // See notes about readLongs above
```

File: lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/packed/LegacyDirectWriter.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+import org.apache.lucene.util.packed.PackedInts;
+
+/**
+ * Class for writing packed integers to be directly read from Directory. Integers can be read
+ * on-the-fly via {@link LegacyDirectReader}.
+ *
+ * <p>Unlike PackedInts, it optimizes for read i/o operations and supports &gt; 2B values. Example
+ * usage:
+ *
+ * <pre class="prettyprint">
```
Comment (dweiss) [inline lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/packed/LegacyDirectWriter.java:32]: You can use code blocks too - then html entities don't need to be escaped..

File: lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/packed/LegacyDirectReader.java [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+      this.in = in;
+      this.offset = offset;
+    }
+
+    @Override
+    public long get(long index) {
+      try {
+        return in.readLong(offset + (index << 3));
+      } catch (IOException e) {
+        throw new RuntimeException(e);
```
Comment (dweiss) [inline lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/packed/LegacyDirectReader.java:364]: UncheckedIOException(e)?

Comment (dweiss) [reply lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/packed/LegacyDirectReader.java:364]: Here and in other places.

**Review by dweiss — COMMENTED** [review]
I peeked at the patch and it looks good to me.

File: lucene/core/src/java/org/apache/lucene/codecs/CodecUtil.java (hunk 1 above)
Comment (iverase) [reply lucene/core/src/java/org/apache/lucene/codecs/CodecUtil.java]: I rename methods like writeBE / readBE to make it more explicit.

File: lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/packed/LegacyDirectWriter.java (hunk 4 above)
Comment (iverase) [reply lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/packed/LegacyDirectWriter.java:32]: This is a copy / paste from DirectWriter so I would prefer to change it in a follow up PR.

File: lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/packed/LegacyDirectReader.java (hunk 5 above)
Comment (iverase) [reply lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/packed/LegacyDirectReader.java:364]: Same as above, maybe a follow up PR? this is already pretty big.

**Comment by iverase** [comment]
I will set the PR ready to review as feedback has been positive so far. I want to stress that the most interesting part is how to deal with reading `segment.gen`. This file does not belong to a codec so we open it blind without knowing the endianness. Therefore the approach I have taken is to write that file always big endian as we are doing until now.

In addition as this files have a codec header / footer, then all headers / footers will be written using BE order. 

Maybe @rmuir and @mikemccand have an opinion here.

**Comment by rmuir** [comment]
> I will set the PR ready to review as feedback has been positive so far. I want to stress that the most interesting part is how to deal with reading `segment.gen`. This file does not belong to a codec so we open it blind without knowing the endianness. Therefore the approach I have taken is to write that file always big endian as we are doing until now.

Agreed, let's try to land the current patch! I think its fine that `segments_N` and codec headers just stay bigendian, as the code for this is nicely self-contained.

**Comment by iverase** [comment]
Thanks @rmuir! I will wait until Monday, if there is no more feedback I will proceed merging the current patch.

**Review by rmuir — APPROVED** [review]
Thanks for all the effort reworking this change: looks great!

**Review by zacharymorn — APPROVED** [review]
Just trying to see how it's being implemented & the changes look great!

**Comment by uschindler** [comment]
Hi @iverase
the merge broke tests: We fixed over the weekend the NRT tests. They had a bug in the security policy, so no NRT Test is running at all. So you did not see the test failure.

If you fix it, make sure to run the replicator tests with -Dtests.nightly=true, so it runs all tests (takes a few minutes)

I will backport the test fixes also to 8.x later. It's great that we fixed it yesterday :-) The NRT tests were not running since now YEARS!

**Comment by uschindler** [comment]
This is great - by the way!

I will make a new PR based on https://github.com/apache/lucene-solr/pull/2176 to check how the little endian by default improves the new MMapDirectory v2 that will hopefully go into Java 17! :-)

**Pull request merged**